### PR TITLE
Add preemption victim selection logic

### PR DESCRIPTION
**Pull request by bsalamat**
**What this PR does / why we need it**:
This is a part of pod preemption work. This PR adds the code to choose minimum set of pods on a node to make enough room for a new higher priority pod.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
The logic to choose one of the possible nodes will be added in a subsequent PR.

**Release note**:

```release-note
NONE
```

ref/ #47604
ref/ #48646

@kubernetes/sig-scheduling-pr-reviews @davidopp

### CODE REVIEW

File: plugin/pkg/scheduler/util/utils.go [hunk 1]
```diff
@@ -39,3 +42,37 @@ func GetUsedPorts(pods ...*v1.Pod) map[int]bool {
 	}
 	return ports
 }
+
+// GetPodPriority return priority of the given pod.
+func GetPodPriority(pod *v1.Pod) int32 {
+	if pod.Spec.Priority != nil {
+		return *pod.Spec.Priority
+	} else {
```
Comment (xiang90) [inline plugin/pkg/scheduler/util/utils.go]: kill this else. already return at line 49

File: plugin/pkg/scheduler/util/utils.go [hunk 2]
```diff
@@ -39,3 +42,37 @@ func GetUsedPorts(pods ...*v1.Pod) map[int]bool {
...
+		// When priority of a running pod is nil, it means it was created at a time
+		// that there was no global default priority class and the priority class
+		// name of the pod was empty. So, we resolve to the static default priority.
+		return scheduling.DefaultPriorityWhenNoDefaultClassExists
+	}
+}
+
+// SortableList is a list that implements sort.Interface.
+type SortableList struct {
```
Comment (xiang90) [inline plugin/pkg/scheduler/util/utils.go]: i feel this does not belong to schedule util but a more generic util pkg.

File: plugin/pkg/scheduler/core/generic_scheduler.go [hunk 3]
```diff
@@ -424,6 +445,74 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
...
+	for _, p := range nodeInfoCopy.Pods() {
+		if util.GetPodPriority(p) < podPriority {
+			podList.Items = append(podList.Items, p)
+			nodeInfoCopy.RemovePod(p)
+		}
+	}
+	// If the new pod does not fit after removing all the lower priority pods,
+	// we are done and this node is not suitable for preemption.
+	if fits, _, _ := podFitsOnNode(pod, meta, nodeInfoCopy, predicates, nil); !fits {
+		return victims, false
```
Comment (xiang90) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: return nil, false

File: plugin/pkg/scheduler/util/utils_test.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import (
+	"sort"
+	"testing"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/kubernetes/pkg/apis/scheduling"
+)
+
+// TestGetPodPriority tests GetPodPriority function.
+func TestGetPodPriority(t *testing.T) {
```
Comment (xiang90) [inline plugin/pkg/scheduler/util/utils_test.go]: can you change this to table driven test?

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 3 above)
Comment (bsalamat) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Done

File: plugin/pkg/scheduler/core/generic_scheduler.go [hunk 5]
```diff
@@ -424,6 +445,74 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
...
+	}
+	// If the new pod fits after the removal of all the lower priority pods, try to
+	// reprieve as may pods as possible starting from the highest priority one.
+	sort.Sort(&podList)
+	for _, p := range podList.Items {
+		lpp := p.(*v1.Pod)
+		nodeInfoCopy.AddPod(lpp)
+		if fits, _, _ := podFitsOnNode(pod, meta, nodeInfoCopy, predicates, nil); !fits {
+			nodeInfoCopy.RemovePod(lpp)
+			victims = append(victims, lpp)
```
Comment (bsalamat) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: The logic looks correct to me. I am not sure what you mean by "remaining", but the loop goes over the whole podList and adds those which make the pod not fit any more are added to `victims`.

File: plugin/pkg/scheduler/util/utils.go (hunk 1 above)
Comment (bsalamat) [inline plugin/pkg/scheduler/util/utils.go]: Done

File: plugin/pkg/scheduler/util/utils_test.go (hunk 4 above)
Comment (bsalamat) [inline plugin/pkg/scheduler/util/utils_test.go]: Done

File: plugin/pkg/scheduler/util/utils.go (hunk 2 above)
Comment (bsalamat) [inline plugin/pkg/scheduler/util/utils.go]: You are probably right, but I don't know any better place, something like a utility library for all components.

**Review by bsalamat — COMMENTED** [review]
Thanks, @xiang90 for your review.

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 5 above)
Comment (xiang90) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: nv. i misread the code.

File: plugin/pkg/scheduler/core/generic_scheduler.go [hunk 6]
```diff
@@ -424,6 +445,74 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
...
+	workqueue.Parallelize(16, len(nodes), checkNode)
+	return nodeNameToPods
+}
+
+// selectVictimsOnNode finds minimum set of pods on the given node that should
+// be preempted in order to make enough room for "pod" to be scheduled.
+// The algorithm first checks if the pod can be scheduled on the node when all the
+// lower priority pods are gone. If so, it sorts all the lower priority pods by
+// their priority and starting from the highest priority one, tries to keep as
+// many of them as possible while checking that the "pod" can still fit on the node.
```
Comment (davidopp) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: This algorithm works well if the infeasibility reason is insufficient resources or pod anti-affinity, but not if the reason is pod affinity. I think maybe for now it's best to just add a comment, as it only affects optimality, not correctness. The observations are:

(1) if the initial infeasibility reasons include pod affinity, then preemption will never help (removing existing pods will never make you satisfy pod affinity if you don't satisfy it initially). So you shouldn't even bother to try preemption when the infeasibility reason is pod affinity
and
(2) if the initial infeasibility reason is something else, say resources, then the short-circuit test ("If the new pod does not fit after removing all the lower priority pods, we are done...") can cause you to exit without trying any reprieving, when in fact reprieving some combination of pods might allow the pod to schedule (in particular, the pod(s) that the new pod requires to satisfy its affinity, as long as it/they can fit resource-wise along with the new pod).

If you wanted to address this, you could

(1) have this function exit immediately if any of the infeasibility reasons are pod affinity
and
(2) at the beginning of this function, check if the pod you are scheduling, or any of the existing pods on the node, have pod affinity, and if so, execute a different version of this function that tries all combinations of existing pods, or does something smarter (there are obviously smarter things you can do)

One other comment: Everything I said above is assuming node-level affinity and anti-affinity. For larger granularity, like rack or zone, I think the deal is
(i) For pod anti-affinity, I think the current algorithm works but is not optimal, in that it will miss an opportunity to preempt on (a) different node(s) than the one you are considering scheduling the pod onto, and thus may say the pod is infeasible everywhere in the cluster when actually it is not (or, will do more or higher-priority preemptions than if it preempted from other node(s)). This is just an optimality issue, so no need to deal with it now.
and
(ii) For pod affinity, I think the rack/zone/whatever case is the same as the single-node case, i.e. what I wrote above for pod affinity still applies.

File: plugin/pkg/scheduler/core/generic_scheduler.go [hunk 7]
```diff
@@ -424,6 +445,74 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
...
+		return nil, false
+	}
+	// If the new pod fits after the removal of all the lower priority pods, try to
+	// reprieve as may pods as possible starting from the highest priority one.
+	podList.Sort()
+	for _, p := range podList.Items {
+		lpp := p.(*v1.Pod)
+		nodeInfoCopy.AddPod(lpp)
+		if fits, _, _ := podFitsOnNode(pod, meta, nodeInfoCopy, predicates, nil); !fits {
+			nodeInfoCopy.RemovePod(lpp)
```
Comment (k82cn) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: If only check it per node, not sure whether it will make preempted pod preempt other lower priority; maybe return all low priority pods. Anyway, let me think about it :).

File: plugin/pkg/scheduler/schedulercache/util.go [hunk 8]
```diff
@@ -27,7 +27,7 @@ func CreateNodeNameToInfoMap(pods []*v1.Pod, nodes []*v1.Node) map[string]*NodeI
 		if _, ok := nodeNameToInfo[nodeName]; !ok {
 			nodeNameToInfo[nodeName] = NewNodeInfo()
 		}
-		nodeNameToInfo[nodeName].addPod(pod)
+		nodeNameToInfo[nodeName].AddPod(pod)
```
Comment (resouer) [inline plugin/pkg/scheduler/schedulercache/util.go]: Why do we need to make cache update functions public here?

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 6 above)
Comment (bsalamat) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: > (1) if the initial infeasibility reasons include pod affinity, then preemption will never help (removing existing pods will never make you satisfy pod affinity if you don't satisfy it initially). So you shouldn't even bother to try preemption when the infeasibility reason is pod affinity

Thanks, @davidopp! Your points are valid, but this function chooses victims for preemption. I haven't "connected" this function yet. Scheduler should never call this function for a pod that cannot be scheduled for pod affinity reasons. Similarly, it shouldn't be called when unschedulability is due to node affinity/anti-affinity. When I connect the function, I will add the checks for these conditions before calling the function

> (2) if the initial infeasibility reason is something else, say resources, then the short-circuit test ("If the new pod does not fit after removing all the lower priority pods, we are done...") can cause you to exit without trying any reprieving, when in fact reprieving some combination of pods might allow the pod to schedule (in particular, the pod(s) that the new pod requires to satisfy its affinity, as long as it/they can fit resource-wise along with the new pod).

This is a valid point. Let me think how I can address it without hurting performance much.

> One other comment: Everything I said above is assuming node-level affinity and anti-affinity. For larger granularity, like rack or zone, I think the deal is
(i) For pod anti-affinity, I think the current algorithm works but is not optimal, in that it will miss an opportunity to preempt on (a) different node(s) than the one you are considering scheduling the pod onto, and thus may say the pod is infeasible everywhere in the cluster when actually it is not (or, will do more or higher-priority preemptions than if it preempted from other node(s)). This is just an optimality issue, so no need to deal with it now.

Yes, but as we discussed, we are not going to address cross node preemption in the first version.

File: plugin/pkg/scheduler/schedulercache/util.go (hunk 8 above)
Comment (bsalamat) [inline plugin/pkg/scheduler/schedulercache/util.go]: In order to use them in scheduler/core.

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 7 above)
Comment (bsalamat) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Sorry, I didn't quite get what you are suggesting here.

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 6 above)
Comment (davidopp) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: BTW I know you have been thinking about how to handle the pod affinity case, but I'm really wondering whether it is necessary for the first version. IIUC, the algorithm you are using will still "work" in the presence of pod affinity, in the same sense that it "works" for anti-affinity when the TopologyKey is larger than a node -- namely, it may say a pod is infeasible everywhere when actually it could be scheduled somewhere (with preemption). But we can explain those circumstances in the release notes and fix it in the next version, I think. (Of course you should make sure this is actually true, and that you can precisely describe the conditions where it will act in this sub-optimal way. Defining those circumstances is a pre-requisite for fixing the issue anyway.)

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 7 above)
Comment (k82cn) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: I'm thinking to return all lower priority pods of this node, and choose pods to preempt cross nodes. Not quick sure whether current solution is OK now, I need to make up some cases here. Did not go through the detail of david's & your comments, but I think there's also question on pod affinity.

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 6 above)
Comment (bsalamat) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Actually, when I was trying to address the pod affinity issue last night, I realized that some of our predicates, including pod affinity/anti-affinity predicates, work with a pod lister which is in actual implementation the SchedulerCache. So, the current logic of adding/removing pods from a NodeInfo does not work correctly for those predicates, as they go to the scheduler cache and find all the potential victim pods still present on the node (despite the fact that `selectVictimsOnNode` removes them from NodeInfo). This seems to be a major problem and may require a completely different solution.

Comment (bsalamat) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: @davidopp I addressed the pod affinity issue. PTAL.

File: plugin/pkg/scheduler/core/generic_scheduler.go [hunk 9]
```diff
@@ -424,6 +445,115 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
...
+		// satisfies the pod affinity rules.
+		for i, p := range potentialVictims.Items {
+			existingPod := p.(*v1.Pod)
+			nodeInfoCopy.AddPod(existingPod)
+			if fits, _, _ = podFitsOnNode(pod, meta, nodeInfoCopy, fitPredicates, nil); !fits {
+				nodeInfoCopy.RemovePod(existingPod)
+			} else {
+				// We found the pod needed to satisfy pod affinity. Let's remove it from
+				// potential victims list.
+				// NOTE: We assume that pod affinity can be satisfied by only one pod,
```
Comment (davidopp) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: There may be more than one pod (already on the node) that satisfies the affinity. (For example, pod affinity rule might be "new pod must run on a node with a pod matching selector S" and this node has several pods matching selector S) You've found and reprieved one of them, but they may not all be the same priority, so a more optimal algorithm would find all of the pods that cause the affinity to be satisfied and then reprieve the highest-priority one. I'm not saying to bother with this optimization, just pointing it out.

File: plugin/pkg/scheduler/core/generic_scheduler.go [hunk 10]
```diff
@@ -424,6 +445,115 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
...
+				potentialVictims.Items = append(potentialVictims.Items[:i], potentialVictims.Items[i+1:]...)
+				break
+			}
+		}
+		if !fits {
+			return nil, false
+		}
+	}
+	victims := []*v1.Pod{}
+	// Try to reprieve as may pods as possible starting from the highest priority one.
```
Comment (davidopp) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Another issue just occurred to me. There may be a pod affinity relationship between two or more of the victims (that is totally unrelated to the pod you are scheduling). Here and elsewhere in this function you are only checking whether the new pod is feasible, not whether the already-running pods' feasibility becomes violated due to preemption. Unfortunately, accounting for this makes this whole process even more computationally expensive (you need to reprieve sets of pods, you can't just greedily reprieve pods).

Maybe for the fist version, ignoring that case is OK. In theory, because we currently only support "RequiredDuringSchedulingIgnoredDuringExecution" affinity we aren't obligated to guarantee pod affinity continues to be met once a pod is scheduled (for example, today if a pod gets evicted for whatever random reason, we don't also evict the other pods on the node that had required pod affinity for the evicted pod -- that's why it's called IgnoredDuringExecution). I say "in theory" because users may be unhappy even if it is "working as intended."

File: plugin/pkg/scheduler/core/generic_scheduler.go [hunk 11]
```diff
@@ -424,6 +445,115 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
...
+		if affinity == nil || affinity.PodAffinity == nil {
+			return nil, false
+		}
+		for _, failedPred := range failedPredicates {
+			if failedPred != predicates.ErrPodAffinityNotMatch {
+				return nil, false
+			}
+		}
+		// If we reach here, it means that the pod cannot be scheduled only due to
+		// pod affinity. Let's try adding pods one at a time and see if any of them
```
Comment (davidopp) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: This is really "pod affinity or pod anti-affinity" right? (since predicates.ErrPodAffinityNotMatch is returned in both cases) I guess this code is still correct because if the failure reason was actually pod anti-affinity, then this is just wasted work and after trying all of the pods you'll find none of them help. In any event it's definitely worth adding a comment about that.

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 9 above)
Comment (bsalamat) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: We sort potential victims by priority and start from the highest priority one. So, the first one that we reprieve is the highest priority one.

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 11 above)
Comment (bsalamat) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: That's correct. At line 515 we check that if we have no pod affinity rules and the failed predicate is affinity, then we don't bother with checking pods as the failure must be related to pod anti-affinity, but if we have both affinity and anti-affinity, we can no longer say which one caused it. So, we may do some extra work.

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 10 above)
Comment (bsalamat) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: This problem exists, but it exists even today as Kubelet may evict a pod due to resource pressure and the eviction could violate affinity rules.
For preemption, it will be computationally very expensive to take care of affinity as topology key may be much broad than a single node, for example, it could be "region". In that case many pods must be checked.
A user can prevent this scenario by having the same priority for pods that have affinity on one another.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 12]
```diff
@@ -984,7 +985,12 @@ func (c *PodAffinityChecker) InterPodAffinityMatches(pod *v1.Pod, meta interface
 	if node == nil {
 		return false, nil, fmt.Errorf("node not found")
 	}
-	if !c.satisfiesExistingPodsAntiAffinity(pod, meta, node) {
+	allPods, err := c.podLister.List(labels.Everything())
```
Comment (ravisantoshgudimetla) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Can we add a metadata precomputation here as this seems to be an expensive operation just like we have metadata for checkServiceAffinity?

Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: @ravisantoshgudimetla we should probably do that, but that change does not belong in this PR. It is an optimization of the predicate function we should be done separately IMO.

Comment (ravisantoshgudimetla) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: @bsalamat Sure, can we have a TODO as part of this PR?

Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: I just checked. `satisfiesExistingPodsAntiAffinity` does use precomputed meta already.

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 9 above)
Comment (davidopp) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Ah, good point.

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 10 above)
Comment (davidopp) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Agreed. I think just adding a comment is sufficient in case someone reading the code wonders about this.

**Comment by bsalamat** [comment]
I noticed that this PR has an issue with `meta` data passed to predicates. Removing/adding pods to NodeInfo requires recomputation of the metadata which can be computationally expensive. Since removals and additions happen frequently, this recomputation could be prohibitively expensive. If you have a good solution for it, please let me know.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 12 above)
Comment (ravisantoshgudimetla) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: So, AFAIU, in `satisfiesExistingPodsAffinity` normal listing is done only when meta is not available but here, we are listing before the call happens to function, so the amortized value may not be equal always.

**Comment by ravisantoshgudimetla** [comment]
>  Removing/adding pods to NodeInfo requires recomputation of the metadata which can be computationally expensive.

@bsalamat  Just curious how you are measuring it(scheduler_perf or any other tool?). I am wondering instead of using AddPod, DeletePod if we invalidate nodeCache by deleting and recreating it(because all of them seem to operations on map) just like (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemon_controller.go#L1194) or use UpdateNodeInfo would be helpful.

**Comment by bsalamat** [comment]
> Just curious how you are measuring it(scheduler_perf or any other tool?). I am wondering instead of using AddPod, DeletePod if we invalidate nodeCache by deleting and recreating it(because all of them seem to operations on map) just like (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemon_controller.go#L1194) or use UpdateNodeInfo would it be helpful?

@ravisantoshgudimetla What daemonset controller does is much simpler and requires much less computation. The `simulate` function tries a single pod on a node and runs only a subset of predicate functions for it. The more complex predicates, such as various types of affinity/anti-affinity are not run by `simulate`. More importantly, preemption needs to remove and add more than one pod to find the minimum set of lower priority pods. If we invalidated nodecache and precomputed meta every time we added/removed a pod on all possible nodes, performance would suffer a lot.

**Comment by bsalamat** [comment]
@davidopp I addressed the issue with updating predicate metadata. This PR is now ready for review, but it depends on my other PR #50805 that @wojtek-t is reviewing and has those changes as well.
New changes in this PR are in the following two commits:
https://github.com/kubernetes/kubernetes/pull/50190/commits/53404967831e5b0dedc114cb8651c8190b80b5a8
https://github.com/kubernetes/kubernetes/pull/50190/commits/190e6ec3fe6764b1f9c4c50081bb1fc003d848a2

**Comment by davidopp** [comment]
Is this still relevant to review given #50949 and #50405?

**Comment by bsalamat** [comment]
@davidopp I sent the a few PRs, the order is #50805, #50190, #50405, #50949. Each one includes all the changes made in a previous PR. So, the last one includes all the changes. I separated them so that we can review and merge them more easily. We can merge #50949 only, but in case of revert, there will be a lot of work for the reviewers again.

**Comment by wojtek-t** [comment]
I agree with @bsalamat - we should merge the PRs one by one. I think this one is first - looking into it now.

**Comment by bsalamat** [comment]
All the changes in this PR are in #50949 as well. So, closing this one.
